### PR TITLE
put all option docs build flavors on equal footing

### PR DESCRIPTION
**Pull request by pennae**
###### Motivation for this change

in #142040 the option docs generation was changed to not sort options.xml in the nix build. in that spirit also stop generating ascii doc and MD doc in nix, and farm that out to scripts too. to make this easier we standardize on options.json as the primary source of options documentation, generating options.xml from options.json when needed. (hopefully docbook will go away soon :crossed_fingers:)

this *may* slightly slow down the system build. the optionsDocBook build is definitely impacted, but the full system build seems to not be (or within the noise level of our measurements).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pennae** [comment]
to set a stage for this a little, this is part of an effort to cache most of the options doc build as well. json is much easier to merge than the xml output nix produces, and caching parts of the doc build requires merging options descriptions. the MD and asciidoc changes happened to line up with the need to make json the primary export format for option docs so they ended up converted as well.

**Review by jtojnar — COMMENTED** [review]
Looks good, just few nits.

File: nixos/doc/manual/default.nix [hunk 1]
```diff
@@ -161,7 +161,7 @@ let
 in rec {
   inherit generatedSources;
 
-  inherit (optionsDoc) optionsJSON optionsXML optionsDocBook;
+  inherit (optionsDoc) optionsAsciiDoc optionsMDDoc optionsJSON optionsDocBook;
```
[the remark this answers is not in the document]
Comment (pennae) [reply nixos/doc/manual/default.nix:164]: might it even make sense to drop asciidoc entirely here? having three two representations of the same data doesn't seem that useful if two of them are just different markup languages

[the remark this answers is not in the document]
Comment (pennae) [reply nixos/doc/manual/default.nix]: we've undone publishing those two attributes for now, but kept the rename in make-options-doc for a later time. we'll have to expand the generator script greatly anyway to get good docs out of it, eg currently it'll generate all examples as a JSON dump of that part of the options set including annotations for generators like `"_type": "literalExample"`

**Review by jtojnar — APPROVED** [review]
LGTM, thanks.

File: nixos/lib/make-options-doc/options-to-docbook.xsl [hunk 2]
```diff
@@ -189,7 +189,7 @@
   </xsl:template>
 
 
-  <xsl:template match="derivation">
+  <xsl:template match="attrs[attr[@name = '_type' and string[@value = 'derivation']]]">
```
Comment (Infinisil) [inline nixos/lib/make-options-doc/options-to-docbook.xsl:192]: What's this change for?

**Review by infinisil — COMMENTED** [review]
This is pretty neat!

File: nixos/lib/make-options-doc/options-to-docbook.xsl (hunk 2 above)
Comment (pennae) [reply nixos/lib/make-options-doc/options-to-docbook.xsl:192]: nix xml makes a distinction between derivations and attrsets, but json can't. to get around this we instead format derivations as a `_type` special value, like options and examples are handled already

Comment (Infinisil) [reply nixos/lib/make-options-doc/options-to-docbook.xsl:192]: Makes sense!

**Review by infinisil — APPROVED** [review]

**Comment by sternenseemann** [comment]
Breaks NixOS with Nix 2.3, see #147265.

**Pull request merged**